drivers: w1: Add driver for Maxim DS2484 Single-Channel 1-Wire Master

casparfriedrich opened the pull request:
This PR adds a family driver for the DS 2482(-100) as well as the DS2484 1-Wire driver. The DS 2482-800 (8-Channel bus master) works as well but the driver is currently lacking multi channel functionality.

casparfriedrich committed:
samples: Add 1-Wire scanner example

This adds a simple example to show basic 1-Wire bus operations

Signed-off-by: Caspar Friedrich <c.s.w.friedrich@gmail.com>

galak commented on drivers/w1/Kconfig.ds248x:
```diff
@@ -0,0 +1,6 @@
+
+config W1_DS248X
+	bool "DS2482/4 Single-Channel 1-Wire Master"
+	depends on I2C
```
Add `default $(dt_compat_enabled,$(DT_COMPAT_MAXIM_DS248X))`

(See `drivers/counter/Kconfig.esp32` for example)

dp7hgh7 replied:
Good idea, done.

galak commented on samples/drivers/w1/scanner/app.overlay:
```diff
@@ -0,0 +1,7 @@
+	w1: w1@18 {
+		compatible = "maxim,ds248x";
+		reg = <0x18>;
+		label = "W1";
```
remove `label` is shouldn't be needed.

dp7hgh7 replied:
`maxim,ds248x.yaml` includes `i2c-device.yaml` which requires it.

galak replied:
ah, ok for now.

galak commented on samples/drivers/w1/scanner/prj.conf:
```diff
@@ -0,0 +1,5 @@
+CONFIG_LOG=y
+CONFIG_SERIAL=y
+CONFIG_W1=y
+CONFIG_W1_DS248X=y
```
remove CONFIG_W1_DS248X, will get set by `default ` in Kconfig change.

dp7hgh7 replied:
Not sure why, but this doesn't work on my machine.

galak replied:
That's probably because you are missing defining `DT_COMPAT_MAXIM_DS248X` in the Kconfig.  If you look at the `drivers/counter/Kconfig.esp32` as an example, you'll see that it has `DT_COMPAT_ESP32_TIMER := espressif,esp32-timer`

galak commented on samples/drivers/w1/scanner/sample.yaml:
```diff
@@ -0,0 +1,2 @@
+sample:
+  name: 1-Wire scanner example
```
Add device tree compat filter check.

`filter: dt_compat_enabled("maxim,ds248x")`

galak commented on samples/drivers/w1/scanner/src/main.c:
```diff
@@ -0,0 +1,26 @@
+
+void main(void)
+{
+	const struct device *dev = DEVICE_DT_GET(DT_NODELABEL(w1));
```
Any reason not to use `DEVICE_DT_GET_ONE()`

dp7hgh7 replied:
Nope, fixed.

galak commented on samples/drivers/w1/scanner/src/main.c:
```diff
@@ -0,0 +1,26 @@
+void main(void)
+{
+	const struct device *dev = DEVICE_DT_GET(DT_NODELABEL(w1));
+	size_t num_devices = w1_search_rom(dev, w1_search_callback, NULL);
```
Add check for device_is_ready()

dp7hgh7 replied:
Done.

galak commented on drivers/w1/w1_ds248x.c:
```diff
@@ -0,0 +1,445 @@
+static int ds248x_init(const struct device *dev)
+{
+	int ret;
+
```
would suggest a local var for `get_config(dev)`

dp7hgh7 replied:
Only in `init(...)`?

galak replied:
Yeah, I was scanning driver from bottom to top.  Personally, I don't think the inline function buys you much w/regards to re-ability vs a local variable for both cfg & data when needed.  (which is what most drivers do).

dp7hgh7 replied:
You mean to have a local variable for every function, right? I think this would add a lot of boilerplate code. I don't have (very) strong feelings about this, I'll give it a try and see if the code becomes to noisy.

galak reviewed (changes requested):
Several DT changes and minor other bits.

galak commented on drivers/w1/Kconfig.ds248x:
```diff
@@ -0,0 +1,7 @@
+# SPDX-License-Identifier: Apache-2.0
+
+config W1_DS248X
+	bool "DS2482/4 Single-Channel 1-Wire Master"
```
You need to add something like:

`DT_COMPAT_MAXIM_DS248X := maxim,ds248x`

dp7hgh7 replied:
I see. I somehow expected that this is something automatic. I wasn't familiar with this structure. Going to integrate this later.

galak replied:
I'm looking to make that a bit more regulator, with https://github.com/zephyrproject-rtos/zephyr/pull/47196, but will see how that progresses.

dp7hgh7 replied:
Good idea! Fixed.

str4t0m reviewed (changes requested):
Thanks for this addition. I had a quick look, and will check the PR in more detail later:

- I believe the sample should not only be runnable with maxim,ds248x, but with all supported  1-wire masters.
You could  for example check with `DT_HAS_COMPAT_STATUS_OKAY` for the zephyr,serial driver and this one, and get the compatible device accordingly.  Similar to how it is done in the `watchdog/wdt_basic_api` test. Subsequently I'd also rename the app.overlay to a generic name such that users can decide which overlay they use.
- Please additionally add a test configuration with this driver to the w1_api test.

casparfriedrich committed:
boards: arm: nrf52840dk_nrf52840: Add missing support tag `arduino_serial`

`arduino_serial` is defined and work but is not listed under supported
tags. Therefore tests depending on this tag are filtered out on this board.

Signed-off-by: Caspar Friedrich <c.s.w.friedrich@gmail.com>

dp7hgh7 commented:
I added some major updates. During the last days I found some details which make the two ICs not fully compatible. Therefore I decided to focus this PR on the DS2484 and provide the DS2482 in a follow up pr. So far I already extracted the common core functionality. Sorry for the inconvenience.

str4t0m reviewed (changes requested):
No problem, it's fine to add the other masters later.
I have added a few comments, please additionally add a test configuration for the new driver to the w1_api test.

dp7hgh7 commented on drivers/w1/w1_ds2484.c:
```diff
@@ -0,0 +1,183 @@
+	uint8_t reg_device_config;
+};
+
+static int ds2484_reset_bus(const struct device *dev)
```
The extracted functions take only the i2c device structure instead of the whole device; therefore they are pretty much needed.

str4t0m replied:
Sorry for the confusion, I implicitly meant if extracting the i2c device in the outer function was the best way to go. Because if you would pass the device you wouldn't need the wrapper for the DS2484, but would still need a wrapper for the channel selection in the case of a ds2482. So yeah, it's probably best to leave it as it is now, and have wrappers for drivers sharing the common ds248x functions.

dp7hgh7 commented on drivers/w1/w1_ds2484.c:
```diff
@@ -0,0 +1,183 @@
+		if (config->slpz_spec.port) {
+			return -ENOTSUP;
+		}
+		if (!device_is_ready(config->slpz_spec.port)) {
```
Done

dp7hgh7 commented on samples/drivers/w1/scanner/src/main.c:
```diff
@@ -0,0 +1,33 @@
+
+	size_t num_devices = w1_search_rom(dev, w1_search_callback, NULL);
+
+	LOG_INF("Number of devices found on bus: %u", num_devices);
```
Done

dp7hgh7 commented on drivers/w1/w1_ds2484.c:
```diff
@@ -0,0 +1,183 @@
+	if (config->slpz_spec.port) {
+		if (!device_is_ready(config->slpz_spec.port)) {
+			LOG_ERR("Port (SLPZ) not ready");
+			return -EINVAL;
```
No strong opinion on that. Done.

dp7hgh7 commented on samples/drivers/w1/scanner/README.rst:
```diff
@@ -0,0 +1,33 @@
+Set DTC_OVERLAY_FILE to "ds2484.overlay" or "serial.overlay" in order to
+enable and configure the drivers.
+
+.. zephyr-app-commands::
```
Took the `serial.overlay` since it's available on pretty much every board.

ycsin commented on drivers/w1/w1_ds248x.h:
```diff
@@ -0,0 +1,121 @@
+extern "C" {
+#endif
+
+int ds248x_reset_bus(const struct i2c_dt_spec *spec);
```
I'd probably document these, but since this is a private header, I'm not sure if it is required

dp7hgh7 replied:
True. Not sure what the usual way is, though.

dp7hgh7 replied:
I think it was about the `extern "C"`-declaration. But I agree, this shouldn't be necessary.

ycsin commented on drivers/w1/w1_ds248x.h:
```diff
@@ -0,0 +1,121 @@
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/zephyr.h>
+
+#define CMD_1WT	 0x78
```
not a blocker but IMO enum is better for these and able to provide editor/compiler warning

str4t0m replied:
Also no strong feelings from my side.

dp7hgh7 replied:
I would leave it for now and revisit for the follow up driver.

galak commented on drivers/w1/CMakeLists.txt:
```diff
@@ -8,6 +8,11 @@ zephyr_library_sources(w1_common.c)
+zephyr_library_sources_ifdef(CONFIG_W1_DS2484         w1_ds2484.c)
+
+if(CONFIG_W1_DS2484)
+	zephyr_library_sources(w1_ds248x.c)
```
why isn't this just `zephyr_library_sources_ifdef(CONFIG_W1_DS2484 w1_ds248x.c)`

dp7hgh7 replied:
Precisely. It should already be in this PR but I found some incompatibilities and didn't want to withdraw the PR altogether. The other drivers will follow.

dp7hgh7 commented on drivers/w1/w1_ds248x.h:
```diff
@@ -0,0 +1,121 @@
+ * Status Register
+ */
+#define STATUS_1WB_pos 0
+#define STATUS_1WB_msk BIT(0)
```
True. Fixed

dp7hgh7 commented on samples/drivers/w1/scanner/ds2484.overlay:
```diff
@@ -0,0 +1,7 @@
+	w1: w1@18 {
+		compatible = "maxim,ds2484";
+		reg = <0x18>;
+		label = "W1";
```
Done.

dp7hgh7 commented on samples/drivers/w1/scanner/serial.overlay:
```diff
@@ -0,0 +1,6 @@
+&arduino_serial {
+	w1: w1 {
+		compatible = "zephyr,w1-serial";
+		label = "W1";
```
done

dp7hgh7 commented on drivers/w1/w1_ds248x.c:
```diff
@@ -0,0 +1,208 @@
+	const uint8_t buf[] = {cmd, data ? *data : 0};
+
+	ret = i2c_write_dt(spec, buf, data ? 2 : 1);
+	if (ret) {
```
ok, fixed.

casparfriedrich committed:
tests: w1: api: Add DS2484 as possible bus master for 1-Wire API tests

Also:
- Restructure test to have different bus master present

Signed-off-by: Caspar Friedrich <c.s.w.friedrich@gmail.com>

dp7hgh7 commented:
> No problem, it's fine to add the other masters later. I have added a few comments, please additionally add a test configuration for the new driver to the w1_api test.

I just added api tests with this driver

str4t0m commented on drivers/w1/Kconfig:
```diff
@@ -24,6 +24,7 @@ config W1_INIT_PRIORITY
 
 rsource "Kconfig.test"
 rsource "Kconfig.zephyr_serial"
+rsource "Kconfig.ds2484"
```
Nit: please sort these three alphabetically.

dp7hgh7 replied:
done

str4t0m commented on drivers/w1/w1_ds248x.c:
```diff
@@ -0,0 +1,200 @@
+static int ds248x_write(const struct i2c_dt_spec *spec, uint8_t cmd, uint8_t *data)
+{
+	int ret;
+
```
No need for an empty line here

dp7hgh7 replied:
I like to keep this helper variable separated from the actual logic. But no super strong opinion here.

str4t0m commented on drivers/w1/w1_ds248x.c:
```diff
@@ -0,0 +1,200 @@
+int ds248x_reset_bus(const struct i2c_dt_spec *spec)
+{
+	int ret;
+
```
same here.

str4t0m commented on drivers/w1/w1_ds248x.c:
```diff
@@ -0,0 +1,200 @@
+int ds248x_write_byte(const struct i2c_dt_spec *spec, uint8_t byte)
+{
+	int ret;
+
```
...

str4t0m commented on drivers/w1/w1_ds248x.c:
```diff
@@ -0,0 +1,200 @@
+		return -EINVAL;
+	}
+
+	uint8_t reg = cfg | ~cfg << 4;
```
move the definition to the beginning of the function.

dp7hgh7 replied:
This assignment is (technically) not always necessary; therefore the early exit was above. Anyway doesn't really make a difference. Done.

str4t0m commented on tests/drivers/w1/w1_api/testcase.yaml:
```diff
@@ -1,9 +1,12 @@
+      fixture: w1_serial_idle
+  drivers.w1.w1-serial:
+    filter: dt_compat_enabled("zephyr,ds2484")
+    platform_allow: nrf52840dk_nrf52840
```
Please review the test case configuration.
It seems as if you would have wanted to add a configuration for the device with extra_args, but only added a filter instead.

dp7hgh7 replied:
Fixed

str4t0m reviewed (changes requested):
Thanks for the update, I have few further comments.

str4t0m commented on samples/drivers/w1/scanner/src/main.c:
```diff
@@ -0,0 +1,33 @@
+
+	int num_devices = w1_search_rom(dev, w1_search_callback, NULL);
+
+	printk("Number of devices found on bus: %u\n", num_devices);
```
Use format specifier for signed integers, or print a different message in case an error is returned.
```suggestion
	printk("Number of devices found on bus: %d\n", num_devices);
```

Also, you could move the definition of num_devices to the beginning of the function.

dp7hgh7 replied:
Format specifier fixed. About the definition: I would like to not split this into more lines than necessary. This variable is not written anywhere else; therefore I don't think this sample would benefit from this.

str4t0m commented on samples/drivers/w1/scanner/sample.yaml:
```diff
@@ -0,0 +1,25 @@
+    extra_args: DTC_OVERLAY_FILE=w1_serial.overlay OVERLAY_CONFIG=w1_serial.conf
+    platform_allow: nrf52840dk_nrf52840
+    harness_config:
+      type: one_line
+      regex:
+        - "Number of devices found on bus: .*"
```
You could move these lines to the common section.

dp7hgh7 replied:
If I add `harness_config` to the common section, the harness_config sections under `tests:` are ignored and therefore the fixtures are not respected.

str4t0m commented on tests/drivers/w1/w1_api/w1_devices.dtsi:
```diff
@@ -0,0 +1,21 @@
+	slave_1: dummy-slave-1 {
+		compatible = "test-w1-dummy-slave";
+		family-code = <0x28>;
+		label = "DUMMY_DEVICE_1";
```
The label can be removed now.

dp7hgh7 replied:
done

galak replied:
`label` is still there :)

str4t0m reviewed (changes requested):
... and some comments regarding the sample and the test.

casparfriedrich committed:
drivers: w1: Add driver for Maxim DS2484 Single-Channel 1-Wire Master

This adds a driver for Maxims DS2484 Single-Channel 1-Wire master
driver. The DS2484 features an extra pin to enable sleep modes which
is available if the pin is configured in the device tree.

Signed-off-by: Caspar Friedrich <c.s.w.friedrich@gmail.com>

dp7hgh7 commented on drivers/w1/Kconfig.ds2484:
```diff
@@ -0,0 +1,9 @@
+# Copyright (c) 2022, Caspar Friedrich <c.s.w.friedrich@gmail.com>
+# SPDX-License-Identifier: Apache-2.0
+
+DT_COMPAT_MAXIM_DS2484 := maxim,ds2484
+
+config W1_DS2484
+	bool "DS2484 Single-Channel 1-Wire Master"
+	depends on I2C
+	default $(dt_compat_enabled,$(DT_COMPAT_MAXIM_DS2484))
```
Done

dp7hgh7 commented on drivers/w1/CMakeLists.txt:
```diff
@@ -8,6 +8,11 @@ zephyr_library_sources(w1_common.c)
 # drivers implementing link functions (read, write, reset)
 zephyr_library_sources_ifdef(CONFIG_W1_TEST           w1_test.c)
 zephyr_library_sources_ifdef(CONFIG_W1_ZEPHYR_SERIAL  w1_zephyr_serial.c)
+zephyr_library_sources_ifdef(CONFIG_W1_DS2484         w1_ds2484.c)
```
Done

dp7hgh7 commented on tests/drivers/w1/w1_api/src/main.c:
```diff
@@ -4,23 +4,17 @@
+#include <zephyr/drivers/w1.h>
+#include <zephyr/zephyr.h>
 
+#include <ztest.h>
```
done

dp7hgh7 commented on samples/drivers/w1/scanner/README.rst:
```diff
@@ -0,0 +1,34 @@
+********************
+
+Set DTC_OVERLAY_FILE to "ds2484.overlay" or "w1_serial.overlay" in order to
+enable and configure the drivers.
```
You mean for the serial driver?

str4t0m replied:
Yes I was referring to the case of the serial overlay specifically.
But we can also add it later if people really forget about that.

dp7hgh7 commented on tests/drivers/w1/w1_api/ds2484.conf:
```diff
@@ -0,0 +1 @@
+CONFIG_I2C=y
```
true. Done.

dp7hgh7 commented on drivers/w1/w1_ds248x.c:
```diff
@@ -0,0 +1,200 @@
+		 * Special value: Don't change read pointer
+		 */
+		break;
+	case REG_PORT:
+	case REG_CONFIG:
+	case REG_CHANNEL:
+	case REG_DATA:
```
Didn't knew that this is a thing in C... Done.

galak commented on drivers/w1/Kconfig.ds2484:
```diff
@@ -0,0 +1,9 @@
+# Copyright (c) 2022, Caspar Friedrich <c.s.w.friedrich@gmail.com>
```
See https://github.com/zephyrproject-rtos/zephyr/pull/48166 on new style for how we are doing Kconfig/DT stuff based on autogenerated Kconfig.dts files (look at <BUILD>/Kconfig/Kconfig.dts)

dp7hgh7 replied:
Fixed

galak commented on tests/drivers/w1/w1_api/w1_devices.dtsi:
```diff
@@ -0,0 +1,21 @@
+	slave_1: dummy-slave-1 {
+		compatible = "test-w1-dummy-slave";
+		family-code = <0x28>;
+		label = "DUMMY_DEVICE_1";
```
`label` prop is still here.

dp7hgh7 replied:
Fixed

galak reviewed (changes requested):
* Kconfig update requested.
* Still need to remove `label` property
* would request `test-w1-dummy-slave.yaml` get updated to: `vnd,w1-device.yaml` (and update `compatible` to match `vnd,w1-device`.

galak commented on drivers/w1/w1_ds2484.c:
```diff
@@ -0,0 +1,181 @@
+	uint8_t reg_device_config;
+};
+
+static int ds2484_reset_bus(const struct device *dev)
```
Should these set of functions be `static inline`?

str4t0m replied:
@galak These are the API functions, do you mean to explicitly inline the ds248x_reset_bus et al. called from here?
If yes, I do agree.

galak replied:
yeah, the ds248x_reset_bus versions.

dp7hgh7 replied:
The header look quite weird then, but impact on build system is reduced this way. Done.

ycsin commented on drivers/w1/Kconfig.ds2484:
```diff
@@ -0,0 +1,8 @@
+config W1_DS2484
+	bool "DS2484 Single-Channel 1-Wire Master"
+	depends on I2C
+	depends on DT_HAS_MAXIM_DS2484_ENABLED
```
@galak should this be like the #47949 as well?

```
depends on I2C && $(dt_compat_on_bus,$(DT_COMPAT_MAXIM_DS2484),i2c)
```

str4t0m replied:
No, that should only be needed if the condition depends on having enabled one of two different buses.
Because there it could happen, that you accidentally enable the wrong bus. I.e. enable I2C but add the device node for SPI.

ycsin replied:
@str4t0m thanks for the enlightenment!

str4t0m commented on drivers/w1/w1_ds248x.h:
```diff
@@ -0,0 +1,302 @@
+
+static inline int ds248x_reset_bus(const struct i2c_dt_spec *spec)
+{
+	int ret;
+
+	uint8_t reg;
```
I don't see the benefit of these empty lines, but won't block on these.

str4t0m commented on drivers/w1/w1_ds2484.c:
```diff
@@ -0,0 +1,181 @@
+
+struct ds2484_config {
+	struct w1_master_config w1_config;
+	const struct i2c_dt_spec i2c_spec;
+	const struct gpio_dt_spec slpz_spec;
```
No `const` is needed in the config struct. (Only if they were pointers to a const struct)
